### ISSUE TITLE
Flag to remove terminal colors and special characters in output

### ISSUE THREAD

**Issue by malav097** [panodata/grafana-wtf#9]
Hi, I am using this tool and needed to export the output to a plain text file to upload it to a Jira issue. The text was bloated with special characters used to give colors to the terminal. I am going to fork this project and work on that feature if you guys don't mind.

**Comment by amotl** [comment]
Dear Simon,

thanks for writing in. Your need absolutely make sense to me, so I created the issue [1] at the upstream `colored` package in order to ask about supporting the `NO_COLOR` environment variable [2]. Let's see what @dslackw will say about this.

In the meanwhile, I would recommend to use a program like `ansifilter` to strip the ANSI control characters used for the output coloring or alternatively use `sed` or `perl` [3].
```
# Remove ANSI characters using "ansifilter"
python -c 'from colored import fg; print(fg("yellow"), "hello", fg("white"))' | ansifilter

# Remove ANSI characters using "sed"
python -c 'from colored import fg; print(fg("yellow"), "hello", fg("white"))' | sed 's/\x1b\[[0-9;]*m//g'

# Remove ANSI characters using "perl"
python -c 'from colored import fg; print(fg("yellow"), "hello", fg("white"))' | perl -pe 's/\x1b\[[^m]+m//g;'
```

What do you think about both ideas?

With kind regards,
Andreas.

[1] https://gitlab.com/dslackw/colored/-/issues/18
[2] https://no-color.org/
[3] https://superuser.com/questions/380772/removing-ansi-color-codes-from-text-stream

**Comment by amotl** [comment]
@Qix- shared some thoughts about the `NO_COLOR` discussion at https://github.com/chalk/supports-color/issues/74#issuecomment-408021917:

> The `NO_COLOR` "manifest" solves nothing that wasn't already available.
> 
> - `FORCE_COLOR=0` for code that uses [supports-color](https://github.com/chalk/supports-color).
> - `<cmd> | cat` for code that checks `isatty()`
> - `TERM=dummy` for code that checks `TERM` and/or uses terminfo/curses (most programs within the last ~40 years)

While adjusting `TERM` within iTerm2 on macOS did not change anything for me, I explored the `isatty()` option.
```bash
$ python -c 'import os, sys; print(os.isatty(sys.stdout.fileno()))'
True

$ python -c 'import os, sys; print(os.isatty(sys.stdout.fileno()))' | cat
False
```

Indeed, both [Rich](https://github.com/willmcgugan/rich) and [colorama](https://github.com/tartley/colorama) already support checking the output stream for being a TTY well and will not emit ANSI control characters when the output stream is redirected to a file or pipe:
- https://github.com/willmcgugan/rich/blob/v9.6.1/rich/console.py#L656-L667
- https://github.com/tartley/colorama/blob/0.4.4/colorama/ansitowin32.py#L43-L53

`NO_COLOR` is already honored by [Rich](https://github.com/willmcgugan/rich), while [colorama](https://github.com/tartley/colorama) is pending.
- https://github.com/willmcgugan/rich/blob/v9.6.1/rich/console.py#L541-L542
- https://github.com/tartley/colorama/pull/230

**Comment by Qix-** [comment]
Chiming in here with two additional _opinions_ on the subject I've formed since that comment:

- Always try to disable the generation of escape codes. Filtering them out afterward is always, _always_ code smell.
- The entire terminal technology scene is archaic and outdated. Very little work is being done to modernize it. If `NO_COLOR` helps people now then fine; the entire scene needs to be reformed in my opinion, so a small additional environment variable isn't the real issue.

If `NO_COLOR` helps people _now_ and it's low cost to implement, then it should be considered. However I still would advise checking for a set of common terminals - xterm-like terminals are best here, e.g. `xterm-256color` - and ignoring any you don't recognize (including `dumb`). Setting your TERM to dumb signals that you want absolutely no control sequences whatsoever (there are more sequences than color codes), which is usually what you want.

Speaking of, the dumb terminal is called `dumb`, not `dummy`. Typo on my part. :)

**Comment by amotl** [comment]
Hi @Qix-,

thank you very much for your general feedback on this. In our specific case, I discovered that the `colored.set_tty_aware()` function has to be called in order to enable checking for `isatty()`, see [1].

@malav097: I will enable that on behalf of the next release of `grafana-wtf`. So, there will be no need to work around that by introducing any other special features to suppress output coloring.

With kind regards,
Andreas.

[1] https://gitlab.com/dslackw/colored/-/issues/18#note_478014411

**Comment by amotl** [comment]
Hi again,

apologies for the long delay. Now, 09fece7 will hopefully fix this problem by using ANSI colors only on TTYs.

With kind regards,
Andreas.

**Comment by amotl** [comment]
Hi once more,

259a73eca is included in version 0.10.0, which just has been published to PyPI. Enjoy!

With kind regards,
Andreas.

P.S.: Feel free to reopen or create another issue if you are still observing problems on this matter.
